core/templates/dev/head/domain/topic/TopicUpdateService.js

```diff
@@ -230,20 +221,58 @@ oppia.factory('TopicUpdateService', [
           // Get the current change list.
           var currentChangeList = UndoRedoService.getChangeList();
           var indicesToDelete = [];
+          for (var i = 0; i < currentChangeList.length; i++) {
+            var changeDict =
+              currentChangeList[i].getBackendChangeObject();
+            if (changeDict.cmd === CMD_MOVE_SKILL_ID_TO_SUBTOPIC) {
+              if (changeDict.new_subtopic_id === subtopicId) {
+                if (changeDict.old_subtopic_id === null) {
+                  indicesToDelete.push(i);
+                } else {
+                  changeDict.cmd = CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC;
+                  changeDict.subtopic_id = changeDict.old_subtopic_id;
+                  delete changeDict.old_subtopic_id;
+                  delete changeDict.new_subtopic_id;
+                }
+              } else if (changeDict.old_subtopic_id === subtopicId) {
+                changeDict.old_subtopic_id = null;
+              }
+            } else if (changeDict.cmd === CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC) {
+              if (changeDict.subtopic_id === subtopicId) {
+                indicesToDelete.push(i);
+              }
+            }
+            currentChangeList[i].setBackendChangeObject(changeDict);
+          }
           for (var i = 0; i < currentChangeList.length; i++) {
             var backendChangeDict =
               currentChangeList[i].getBackendChangeObject();
-            if (backendChangeDict.subtopic_id === subtopicId) {
-              // The indices in the change list corresponding to changes to the
-              // currently deleted and newly created subtopic are to be removed
-              // from the list.
-              indicesToDelete.push(i);
+            if (indicesToDelete.indexOf(i) !== -1) {
               continue;
             }
-            // When a newly created subtopic is deleted, the subtopics created
-            // after it would have their id reduced by 1.
-            if (backendChangeDict.subtopic_id > subtopicId) {
-              backendChangeDict.subtopic_id--;
+            if (backendChangeDict.hasOwnProperty('subtopic_id')) {
+              if (backendChangeDict.subtopic_id === subtopicId) {
+                // The indices in the change list corresponding to changes to
+                // the currently deleted and newly created subtopic are to be
+                // removed from the list.
+                indicesToDelete.push(i);
```

Comment:
Woah, something's off here. You shouldn't be modifying indicesToDelete in the same pass as when you're depending on it to tell you what to use and not use. Why not do this part in the previous loop?

Reply:
That initial check is only to skip over those that were to be deleted after the previous loop through changelists regarding the move operation. Actually, that check is not required, since even though this loop does modify those change dicts, it will be removed anyways. Will remove that check.

Reply:
Done.